warehouse/query-core/src/main/java/org/apache/commons/jexl3/parser/JexlNodes.java

```diff
@@ -195,25 +193,113 @@ public static <T extends JexlNode> T swap(T parent, JexlNode a, JexlNode b) {
     public static JexlNode promote(JexlNode parent, JexlNode child) {
         JexlNode grandpa = parent.jjtGetParent();
         if (grandpa == null) {
-            child.parent = null;
+            child.jjtSetParent(null);
             return child;
         } else {
             return swap(parent.jjtGetParent(), parent, child);
         }
     }
 
-    public static void setLiteral(ASTNumberLiteral literal, Number value) {
+    public static boolean setLiteral(ASTNumberLiteral literal, Number value) {
+        boolean success = false;
+
         Preconditions.checkNotNull(literal);
         Preconditions.checkNotNull(value);
 
-        literal.literal = value;
+        if (isNaturalNumber(value.getClass())) {
+            literal.setNatural(value.toString());
+            success = true;
+        } else if (isRealNumber(value.getClass())) {
+            literal.setReal(value.toString());
+            success = true;
+        }
+
+        return success;
+    }
+
+    protected static boolean isNaturalNumber(Class<? extends Number> clazz) {
+        return clazz == Integer.class || clazz == Long.class || clazz == BigInteger.class;
+    }
+
+    protected static boolean isRealNumber(Class<? extends Number> clazz) {
+        return clazz == Float.class || clazz == Double.class || clazz == BigDecimal.class;
     }
 
     public static void setLiteral(ASTStringLiteral literal, String value) {
         Preconditions.checkNotNull(literal);
         Preconditions.checkNotNull(value);
 
-        literal.image = value;
+        literal.setLiteral(value);
+    }
+
+    public static void setLiteral(ASTRegexLiteral literal, String value) {
+        Preconditions.checkNotNull(literal);
+        Preconditions.checkNotNull(value);
+
+        literal.setLiteral(value);
+    }
+
+    public static void setLiteral(ASTJxltLiteral literal, String value) {
+        Preconditions.checkNotNull(literal);
+        Preconditions.checkNotNull(value);
+
+        literal.setLiteral(value);
+    }
+
+    public static void setAnnotation(ASTAnnotation annotation, String name) {
+        Preconditions.checkNotNull(annotation);
+        Preconditions.checkNotNull(name);
+
+        annotation.setName(name);
+    }
+
+    public static void setQualifiedIdentifier(ASTQualifiedIdentifier identifier, String value) {
+        Preconditions.checkNotNull(identifier);
+        Preconditions.checkNotNull(value);
+
+        identifier.setName(value);
+    }
+
+    public static void setIdentifier(ASTIdentifier identifier, String name) {
+        Preconditions.checkNotNull(identifier);
+        Preconditions.checkNotNull(name);
+
+        identifier.setSymbol(name);
+    }
+
+    public static void setIdentifierAccess(ASTIdentifierAccess identifierAccess, String value) {
+        Preconditions.checkNotNull(identifierAccess);
+        Preconditions.checkNotNull(value);
+
+        identifierAccess.setIdentifier(value);
+    }
+
+    public static Object getImage(JexlNode node) {
```

Comment:
I think we may need to create a getStringImage(JexlNode node) that returns the string version instead of using String.valueOf everywhere.  This way we can also handle returning a null image as null instead of the string "null".

Reply:
check out JexlNodes.getIdentifierOrLiteralAsString.